httpcore/_synchronization.py

```diff
@@ -112,117 +128,95 @@ def __exit__(
+            self._semaphore = trio.Semaphore(
                 initial_value=self._bound, max_value=self._bound
             )
+        else:
+            # Note: asyncio.BoundedSemaphore has better performance characteristics than anyio.Semaphore
+            self._semaphore = asyncio.BoundedSemaphore(self._bound)
 
     async def acquire(self) -> None:
-        if not self._backend:
+        if self._semaphore is None:
             self.setup()
-
-        if self._backend == "trio":
-            await self._trio_semaphore.acquire()
-        elif self._backend == "asyncio":
-            await self._anyio_semaphore.acquire()
+        semaphore: _LockProto = self._semaphore  # type: ignore[assignment]
+        await semaphore.acquire()
 
     async def release(self) -> None:
-        if self._backend == "trio":
-            self._trio_semaphore.release()
-        elif self._backend == "asyncio":
-            self._anyio_semaphore.release()
+        semaphore: _LockProto = self._semaphore  # type: ignore[assignment]
+        semaphore.release()
 
 
-class AsyncShieldCancellation:
-    # For certain portions of our codebase where we're dealing with
-    # closing connections during exception handling we want to shield
-    # the operation from being cancelled.
-    #
-    # with AsyncShieldCancellation():
-    #     ... # clean-up operations, shielded from cancellation.
-
-    def __init__(self) -> None:
-        """
-        Detect if we're running under 'asyncio' or 'trio' and create
-        a shielded scope with the correct implementation.
-        """
-        self._backend = current_async_library()
-
-        if self._backend == "trio":
-            self._trio_shield = trio.CancelScope(shield=True)
-        elif self._backend == "asyncio":
-            self._anyio_shield = anyio.CancelScope(shield=True)
-
-    def __enter__(self) -> "AsyncShieldCancellation":
-        if self._backend == "trio":
-            self._trio_shield.__enter__()
-        elif self._backend == "asyncio":
-            self._anyio_shield.__enter__()
-        return self
-
-    def __exit__(
-        self,
-        exc_type: Optional[Type[BaseException]] = None,
-        exc_value: Optional[BaseException] = None,
-        traceback: Optional[TracebackType] = None,
-    ) -> None:
-        if self._backend == "trio":
-            self._trio_shield.__exit__(exc_type, exc_value, traceback)
-        elif self._backend == "asyncio":
-            self._anyio_shield.__exit__(exc_type, exc_value, traceback)
+async def async_cancel_shield(
+    shielded: Callable[[], Coroutine[Any, Any, None]],
+) -> None:
+    if current_async_backend() == "trio":
+        with trio.CancelScope(shield=True):
+            await shielded()
+    else:
+        inner_task = asyncio.create_task(shielded())
+        retry = False
+        while True:
+            try:
+                await asyncio.shield(inner_task)
+                break
+            except asyncio.CancelledError:
```

Comment:
Why are we not just using a standard shield here?

```python
inner_task = asyncio.create_task(shielded())
await asyncio.shield(inner_task)
```

Reply:
Its to wait for the `inner_task` to finish (unless something is repeatedly trying to cancel this). But if we dont need to wait for the given task to finish then it is not needed and we can use just the simple shield